### PR TITLE
Leave strain processing loop early if peak is zero

### PR DESCRIPTION
**Pull request by MaxOhn**
In a similar vein as #18067, these changes go a step further and avoid unnecessary iterations and stored strains altogether.

If `currentSectionPeak` is already zero it can't decay any further so all that's happening is redundant zeros are stored in the `strainPeaks` list and `currentSectionEnd` is incremented.

This greatly improves performance for edge cases such as [/b/3739922](https://osu.ppy.sh/beatmapsets/1392663#osu/3739922) and in general all maps with zero strain sections should benefit.

As I'm not particularly familiar with C# tooling I can't provide any specific benchmark values unfortunately.

I'm in the process of calculating current difficulty and performance values for ~100k maps through osu-tools and will compare them with a [ported version of this change](https://github.com/MaxOhn/rosu-pp/blob/7cc3a4644b62116a1838b3c307290e6fe8780edd/src/osu/skills/traits.rs#L55-L61) to see if they still coincide. Having someone else perform a similar check would be appreciated, just to be sure.

### CODE REVIEW

File: osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs [hunk 1]
```diff
@@ -57,6 +58,14 @@ public sealed override void Process(DifficultyHitObject current)
                 saveCurrentPeak();
                 startNewSectionFrom(currentSectionEnd, current);
                 currentSectionEnd += SectionLength;
+
+                // If the current peak is 0 it can't decay any further so we can leave early
+                if (Precision.AlmostEquals(currentSectionPeak, 0) && current.StartTime > currentSectionEnd)
```
Comment (MaxOhn) [inline osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs:63]: Unsure about the preferred acceptable difference for double comparison so I've kept it at the default for now.

**Comment by MaxOhn** [comment]
I forgot that the strain peak lists are required to have the same size in taiko which makes this optimization incorrect.

As alternative one could 
- call `saveCurrentPeak()` `remainingIterations` many times so that the zeros are still stored but at least the iteration itself is still simplified since `currentSectionEnd` isn't checked every time
- store `strainPeaks` in a wrapper type representing a list that stores elements which are either non-zero doubles or an integer representing an amount of zeros

The former is simple but doesn't have as much of a performance gain. Also the issue of gigantic strain lists for abstruse edge case maps that are really long, especially on a really low clock rate, wouldn't be solved.

The latter would throw a little more complexity into things and I'm not sure if that would be welcome. It could potentially also lead to a minor performance regression on maps that wouldn't sufficiently benefit from it. I'm not sure about C#'s performance difference for lists of doubles compared to a custom list wrapper as described.

**Comment by peppy** [comment]
Good rule of thumb to run tests locally 😅

**Comment by MaxOhn** [comment]
> Good rule of thumb to run tests locally 😅

My apologies, I went a little to fast 🥴 

I adjusted the initial implementation so that zero strain values are still stored, just in a compact manner. The resulting difficulty values should therefore trivially be the same.

As expected, the edge case map [/b/3739922](https://osu.ppy.sh/beatmapsets/1392663#osu/3739922) now performs much better. On regular maps without (major) zero strain sections I couldn't see any significant loss in performance but I only did a very rudimentary benchmark on a handful of maps.

**Review by adamLutPhi — COMMENTED** [review]
Changes been reviewed.

**Comment by peppy** [comment]
Rather than the complexity you have, would it not be equally beneficial (and simpler) to always store adjacent values using the `value`+`count` method? Ie. removing the generic `Number` and always appending to the previous number if it's the same strain value?

**Comment by MaxOhn** [comment]
I did consider it and would've done that immediately if I was dealing with integers but for floats I was worried that it would introduce too many inaccuracies since it relies heavier on element comparison.

The current state seemed to result in the exact same difficulty attributes when I ran it over a ton of maps (albeit osu!standard-only).
I'll implement your suggestion and see how much the acceptable difference during float comparison impacts the results.

**Comment by MaxOhn** [comment]
After fiddling around with the acceptable difference for float precision, turns out the default `1E-7` and even `1E-8` result in values that are off ever so slightly but with `1E-9` everything (~100k osu!std, 15k osu!mania) seems to match so I've put it on `1E-10` for now.

File: osu.Game/Utils/CompactList.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// Adds a value to the <see cref="CompactList"/>.
+        /// If <paramref name="value"/> equals the current last value, it will be stored compactly.
+        /// </summary>
+        /// <param name="value">The value to be added to the list.</param>
+        /// <param name="count">The amount of times <paramref name="value"/> is added to the list.</param>
+        public void Add(double value, int count = 1)
+        {
+            if (items.LastOrDefault() is Number number && Precision.AlmostEquals(number.Value, value, 1E-10))
```
Comment (peppy) [inline osu.Game/Utils/CompactList.cs:40]: You'll definitely want to explain how/where you came up with this precision value if we go with this solution.

**Comment by MaxOhn** [comment]
Checked the acceptable difference some more and compared to osu-tools itself instead of a port and even found a case where machine epsilon was still too large as it was causing a tiny difference.
With `1E-16` I couldn't find differences anymore but that said, I don't have the means to run osu-tools over all maps so I only checked for 1k random maps of each ruleset with 3-5 different mod combinations.

Since these changes conceptually don't influence any logic, if any potential remaining inaccuracies are still of concern, it might be an option to consider adopting these changes with the next update to difficulty attributes.

File: osu.Game/Utils/CompactList.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            return GetEnumerator();
+        }
+    }
+
+    internal class Number
+    {
+        public double Value { get; private set; }
+
+        public double Count { get; set; }
```
Comment (konvovden) [inline osu.Game/Utils/CompactList.cs:79]: Is it correct to use here `double` rather than `int`?

Comment (MaxOhn) [reply osu.Game/Utils/CompactList.cs:79]: Good catch, `int` definitely makes more sense here

**Comment by peppy** [comment]
Just for sanity I've also added benchmarks on an arbitrary beatmap (non-outlier) so we can get a feel for how this affects performance:

master:

|                              Method |      Mean |     Error |    StdDev |     Gen0 |     Gen1 |    Gen2 | Allocated |
|------------------------------------ |----------:|----------:|----------:|---------:|---------:|--------:|----------:|
|   BenchmarkOsuDifficultyCalculation | 12.390 ms | 0.1596 ms | 0.1415 ms | 296.8750 | 156.2500 | 31.2500 |   3.13 MB |
| BenchmarkTaikoDifficultyCalculation |  3.752 ms | 0.0492 ms | 0.0436 ms | 195.3125 |  97.6563 |       - |   2.16 MB |
| BenchmarkCatchDifficultyCalculation |  9.300 ms | 0.1534 ms | 0.1360 ms | 359.3750 | 187.5000 | 46.8750 |   3.76 MB |
| BenchmarkManiaDifficultyCalculation |  4.019 ms | 0.0627 ms | 0.0587 ms | 203.1250 | 101.5625 |       - |   2.24 MB |

this pr:

|                              Method |      Mean |     Error |    StdDev |     Gen0 |     Gen1 |    Gen2 | Allocated |
|------------------------------------ |----------:|----------:|----------:|---------:|---------:|--------:|----------:|
|   BenchmarkOsuDifficultyCalculation | 12.060 ms | 0.1870 ms | 0.1562 ms | 296.8750 | 156.2500 | 31.2500 |   3.17 MB |
| BenchmarkTaikoDifficultyCalculation |  3.766 ms | 0.0631 ms | 0.0590 ms | 203.1250 | 101.5625 |       - |   2.22 MB |
| BenchmarkCatchDifficultyCalculation |  9.646 ms | 0.1825 ms | 0.2998 ms | 375.0000 | 187.5000 | 46.8750 |   3.77 MB |
| BenchmarkManiaDifficultyCalculation |  3.960 ms | 0.0694 ms | 0.0800 ms | 203.1250 | 101.5625 |       - |   2.26 MB |

**Review by peppy — APPROVED** [review]

**Comment by smoogipoo** [comment]
osu: https://docs.google.com/spreadsheets/d/1qtGC8uiqsKYpkanLmRyalHn0Aowrbixbf7s0k8yfeF0/edit
taiko: TBD
catch: TBD
mania: TBD

**Comment by peppy** [comment]
Given there looks to be issues on some rulesets, and the average map is either same or negligibly worse performance, I'm not sure how valid this change is. Would want to see a large change in performance of a larger batch run to consider this further (as it is increasing the complexity of the code).

**Comment by MaxOhn** [comment]
The performance difference is indeed not quite what I was hoping for.
One other big reason for this PR is that an application of mine, which uses a port of the difficulty calculation, would crash due to memory allocation on these niche maps with super low clock rates since the strain lists directly depend on these.
Without this PR, clock rates from user input should probably always be sanitized to not be too low (which lazer presumably already does but it highlights the importance).
I would agree though that the current benefit doesn't quite outweigh the added complexity.
The surprisingly large differences in PP are pretty curious, could look into fixing these specific cases but I'm not sure it would justify the PR going through.

**Comment by peppy** [comment]
Going to close this for now as stalled / not required.

**Pull request closed without merging**